### PR TITLE
Network Server deduplication, uplink testing and fixes

### PR DESCRIPTION
**Pull request by rvolosatovs**
#### Summary

Closes https://github.com/TheThingsIndustries/lorawan-stack/issues/1985
Refs https://github.com/TheThingsIndustries/lorawan-stack-aws/issues/113

#### Changes

- Deduplicate uplinks after matching the device (and context)
- Reimplement uplink testing
- Implement deduplication in Redis
- Add testing utilities
- Consider frequency and data rate when deduplicating uplink. (If different settings are used, the uplink is not a duplicate - it's a retransmission, which we may be dropping this way)
- Do not wait for collection window to be finished before returning in `GsNs.HandleUplink`
- Fix various uplink issues discovered as a consequence of uplink testing:
   - Always match the lowest data rate index when searching data rate (see `NOTE`)
   - Fix ADR uplink handling (currently we store at most 1)
   - Fix data uplink drop event handling
   - Fix retransmission-related event and metric handling
   - Return proper TTN error on rejoins

#### Notes for reviews

Network Server uplink tests now contain functionality to test all available MAC/PHY/FP combinations. Such tests take enormous amounts of time(minutes), so by default NS tests combinations of:
FP: EU868, US915
MAC: 1.1, 1.0.4, 1.0.3
PHY: 1.1-b, 1.0.3-a, 1.0.2-b

If `-short` is specified:
FP: EU868, US915
MAC: 1.0.4
PHY:  1.0.3-a

Setting `prio/medium` due to https://github.com/TheThingsIndustries/lorawan-stack/issues/1985

Note, that the large amount of changes is due to `GsNs` testing changes 
2020-03-10-12:12:27-screenshot

Please review commit-by-commit. @htdvisser just CLI and utils is enough

Many of the changes cannot be extracted due to the changes in testing unfortunately. Some small 1-2 line change commits (i.e. not worth a separate PR) can be though, let me know if you want me to do that.

#### Checklist

- [ ] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/util/test/time.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+// MockClock is used to mock time package functionality.
+type MockClock struct {
+	nowMu sync.RWMutex
+	now   time.Time
+
+	nowChsMu sync.RWMutex
+	nowChs   map[chan<- time.Time]struct{}
+}
+
+func NewMockClock(t time.Time) *MockClock {
```
Comment (htdvisser) [inline pkg/util/test/time.go]: Missing Godoc.

File: pkg/band/band.go [hunk 2]
```diff
@@ -305,10 +305,16 @@ func (b Band) FindSubBand(frequency uint64) (SubBandParameters, bool) {
 	return SubBandParameters{}, false
 }
 
-// FindDataRate returns the data rate with index by API data rate, if any.
+// FindDataRate returns the uplink data rate with index by API data rate, if any.
 func (b Band) FindDataRate(dr ttnpb.DataRate) (ttnpb.DataRateIndex, DataRate, bool) {
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply pkg/band/band.go]: I checked usages and it seemed that this method is only used for uplinks

File: pkg/networkserver/redis/uplink_deduplicator.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	_, err := up.Settings.DataRate.Modulation.MarshalTo(drBytes)
+	if err != nil {
+		return "", err
+	}
+	h := fnv.New64a()
+	_, _ = h.Write(up.RawPayload)
+	return ttnredis.Key(
+		keyEncoding.EncodeToString(h.Sum(nil)),
+		strconv.FormatUint(up.Settings.Frequency, 32),
+		keyEncoding.EncodeToString(drBytes),
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply pkg/networkserver/redis/uplink_deduplicator.go]: Retransmission support - frames transmitted on a different channel/data rate are not duplicates(i.e. the same message received by different gateways), those are different transmissions

File: cmd/ttn-lw-stack/commands/start.go [hunk 4]
```diff
@@ -167,6 +167,9 @@ var startCommand = &cobra.Command{
 			config.NS.Devices = &nsredis.DeviceRegistry{
 				Redis: redis.New(config.Redis.WithNamespace("ns", "devices")),
 			}
+			config.NS.UplinkDeduplicator = &nsredis.UplinkDeduplicator{
+				Redis: redis.New(config.Redis.WithNamespace("ns", "uplink-deduplication")),
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply cmd/ttn-lw-stack/commands/start.go]: Was not aware we have 2 :thinking:

File: pkg/networkserver/redis/uplink_deduplicator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (htdvisser) [reply pkg/networkserver/redis/uplink_deduplicator.go]: Ok, maybe good to add a code comment to explain that.

File: pkg/band/band.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (htdvisser) [reply pkg/band/band.go]: Fine, but I still think it should be renamed. Small effort to prevent future bugs.

**Pull request merged**